server/packetio.go

```diff
@@ -160,42 +213,164 @@ func (p *packetIO) writePacket(data []byte) error {
 	length := len(data) - 4
 	server_metrics.WritePacketBytes.Add(float64(len(data)))
 
-	for length >= mysql.MaxPayloadLen {
+	maxPayloadLen := mysql.MaxPayloadLen
+	if p.compressionAlgorithm != mysql.CompressionNone {
+		maxPayloadLen -= 4
+	}
+
+	for length >= maxPayloadLen {
 		data[3] = p.sequence
 		data[0] = 0xff
 		data[1] = 0xff
 		data[2] = 0xff
 
-		if n, err := p.bufWriter.Write(data[:4+mysql.MaxPayloadLen]); err != nil {
-			return errors.Trace(mysql.ErrBadConn)
-		} else if n != (4 + mysql.MaxPayloadLen) {
-			return errors.Trace(mysql.ErrBadConn)
+		if p.compressionAlgorithm != mysql.CompressionNone {
+			if n, err := p.compressedWriter.Write(data[:4+maxPayloadLen]); err != nil {
```

Comment:
seems we only use `Write` and `Flush` for both `compressedWriter` and `bufWriter`, maybe we can abstract an interface so there's no if-branch

Reply:
@dveeden I have no strong opinion about it. You can merge this PR or address it in following commit.